cli/util/util.go

```diff
@@ -91,3 +140,137 @@ func GetHelpCommand(cmd *cobra.Command) *cobra.Command {
 
 	return nil
 }
+
+// GetHomeDir returns current home directory.
+func GetHomeDir() (string, error) {
+	usr, err := user.Current()
+	if err != nil {
+		return "", err
+	}
+	return usr.HomeDir, nil
+}
+
+func readFromPos(f *os.File, pos int64, buf *[]byte) (int, error) {
+	if _, err := f.Seek(pos, io.SeekStart); err != nil {
+		return 0, fmt.Errorf("Failed to seek: %s", err)
+	}
+
+	n, err := f.Read(*buf)
+	if err != nil {
+		return n, fmt.Errorf("Failed to read: %s", err)
+	}
+
+	return n, nil
+}
+
+// GetLastNLinesBegin return the position of last lines begin.
+func GetLastNLinesBegin(filepath string, lines int) (int64, error) {
+	if lines == 0 {
+		return 0, nil
+	}
+
+	if lines < 0 {
+		lines = -lines
+	}
+
+	f, err := os.Open(filepath)
+	if err != nil {
+		return 0, fmt.Errorf("Failed to open file: %s", err)
+	}
+	defer f.Close()
+
+	var fileSize int64
+	if fileInfo, err := os.Stat(filepath); err != nil {
+		return 0, fmt.Errorf("Failed to get fileinfo: %s", err)
+	} else {
+		fileSize = fileInfo.Size()
+	}
+
+	if fileSize == 0 {
+		return 0, nil
+	}
+
+	buf := make([]byte, bufSize)
+
+	var filePos int64 = fileSize - bufSize
+	var lastNewLinePos int64 = 0
+	var newLinesN int = 0
+
+	// check last symbol of the last line
+
+	if _, err := readFromPos(f, fileSize-1, &buf); err != nil {
+		return 0, err
+	}
+	if buf[0] != '\n' {
+		newLinesN++
+	}
+
+	lastPart := false
+
+Loop:
+	for {
+		if filePos < 0 {
+			filePos = 0
+			lastPart = true
+		}
+
+		n, err := readFromPos(f, filePos, &buf)
+		if err != nil {
+			return 0, err
+		}
+
+		for i := n - 1; i >= 0; i-- {
+			b := buf[i]
+
+			if b == '\n' {
+				newLinesN++
+			}
+
+			if newLinesN == lines+1 {
+				lastNewLinePos = filePos + int64(i+1)
+				break Loop
+			}
+		}
+
+		if lastPart || filePos == 0 {
+			break
+		}
+
+		filePos -= bufSize
+	}
+
+	return lastNewLinePos, nil
+}
+
+func GetLastNLines(filepath string, linesN int) ([]string, error) {
+	lastNLinesBeginPos, err := GetLastNLinesBegin(filepath, linesN)
+	if err != nil {
+		return nil, err
+	}
+
+	file, err := os.Open(filepath)
```

Comment:
Why not using mmap technique for working with files. IMHO it will be simpler.

Reply:
This is a good suggestion, I will implement it as part of the "folow up".